### ISSUE TITLE
HTTP_PROTOTYPE is a pain

### ISSUE THREAD

**Issue by davidgfnet** [oktal/pistache#134]
Hey there, I was trying to start a very simple server that can be instantiated with some arguments, ie:

```
Pistache::Http::Endpoint endpoint("*:9999");
endpoint.init(Pistache::Http::Endpoint::options().threads(4));
endpoint.setHandler(Pistache::Http::make_handler<WebHandler>(123));
endpoint.serve();
```

The main issue is that the clone() function is being implemented by calling the default constructor, which is wrong I guess?
So obvious next step: implement your clone(). That's a pain since HTTP_PROTOTYPE essentially includes an implementation of that function and you can't re-define it. Unless of course, you use two levels of subclassing, like:

```
struct WebHandlerBase : public Pistache::Http::Handler {
	HTTP_PROTOTYPE(WebHandlerBase)
	void onRequest(const Pistache::Http::Request& req, Pistache::Http::ResponseWriter response) {}
};

struct WebHandler : public WebHandlerBase {
	void onRequest(const Pistache::Http::Request& req, Pistache::Http::ResponseWriter response) {
		// STUFF HERE
	}

	std::shared_ptr<Pistache::Tcp::Handler> clone() const {
		return Pistache::Http::make_handler<WebHandler>(this);
	}

	WebHandler(const WebHandler *other) : f(other->f) {}
	WebHandler(int arg1) : attr1(arg1) {}

	int attr1;
};
```

Now, how would you do this without using two classes but just one?
Thanks for the great library!

**Comment by jgh-** [comment]
Yeah the clone function was a bit of a head scratcher for me considering the server is passed around via `shared_ptr` and presumably you want to know which instance is being used.

**Comment by davidgfnet** [comment]
You can just define clone as a pure virtual function and provide a macro that defines a trivial implementation for it. I wouldnt add a trivial impl. in the upper class cause that way (assuming the trivial implemnentation uses the default constructor) you are forcing a class to have a default constructor. Or for what it matters you can use the copy constructor, which I think is ok, since it is always a good idea to define a proper copy constructor for your classes anyway.

**Comment by jgh-** [comment]
IIRC that is how it works right now, but I don't think it uses the copy constructor.

**Comment by hoang408** [comment]
@davidgfnet , could you please explain `WebHandler(const WebHandler *other) : f(other->f) {}`?  What is `f` that you are referring to?  I dont see it defined anywhere.  Thanks.
